File: app/csharp/AccountingSystem.Api/Infrastructure/Persistence/Repositories/AccountRepository.cs
using AccountingSystem.Application.Ports.Out;
using AccountingSystem.Infrastructure.Persistence.Dapper.Entities;
using Dapper;
using Npgsql;

namespace AccountingSystem.Infrastructure.Persistence.Repositories;

/// <summary>
/// 勘定科目リポジトリ
/// </summary>
public class AccountRepository : IAccountRepository
{
    private readonly string _connectionString;

    // 共通のSELECT句（ENUM型を文字列にキャスト）
    private const string SelectColumns = @"
        ""勘定科目ID"",
        ""勘定科目コード"",
        ""勘定科目名"",
        ""勘定科目カナ"",
        ""勘定科目種別""::text AS ""勘定科目種別"",
        ""合計科目"",
        ""BSPL区分"",
        ""取引要素区分"",
        ""費用区分"",
        ""表示順序"",
        ""集計対象"",
        ""課税取引コード"",
        ""残高"",
        ""作成日時"",
        ""更新日時""
    ";

    static AccountRepository()
    {
        // Dapper のカスタムプロパティマッピング設定
        SqlMapper.SetTypeMap(
            typeof(Account),
            new CustomPropertyTypeMap(
                typeof(Account),
                (type, columnName) => columnName switch
                {
                    "勘定科目ID" => type.GetProperty(nameof(Account.AccountId)),
                    "勘定科目コード" => type.GetProperty(nameof(Account.AccountCode)),
                    "勘定科目名" => type.GetProperty(nameof(Account.AccountName)),
                    "勘定科目カナ" => type.GetProperty(nameof(Account.AccountNameKana)),
                    "勘定科目種別" => type.GetProperty(nameof(Account.AccountType)),
                    "合計科目" => type.GetProperty(nameof(Account.IsSummaryAccount)),
                    "BSPL区分" => type.GetProperty(nameof(Account.BsplType)),
                    "取引要素区分" => type.GetProperty(nameof(Account.TransactionElementType)),
                    "費用区分" => type.GetProperty(nameof(Account.ExpenseType)),
                    "表示順序" => type.GetProperty(nameof(Account.DisplayOrder)),
                    "集計対象" => type.GetProperty(nameof(Account.IsAggregationTarget)),
                    "課税取引コード" => type.GetProperty(nameof(Account.TaxCode)),
                    "残高" => type.GetProperty(nameof(Account.Balance)),
                    "作成日時" => type.GetProperty(nameof(Account.CreatedAt)),
                    "更新日時" => type.GetProperty(nameof(Account.UpdatedAt)),
                    _ => null
                }
            )
        );
    }

    public AccountRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    public async Task<int> InsertAsync(Account account)
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = @"
            INSERT INTO ""勘定科目マスタ"" (
                ""勘定科目コード"",
                ""勘定科目名"",
                ""勘定科目カナ"",
                ""勘定科目種別"",
                ""合計科目"",
                ""BSPL区分"",
                ""取引要素区分"",
                ""費用区分"",
                ""表示順序"",
                ""集計対象"",
                ""課税取引コード"",
                ""残高"",
                ""作成日時"",
                ""更新日時""
            ) VALUES (
                @AccountCode,
                @AccountName,
                @AccountNameKana,
                @AccountType::account_type,
                @IsSummaryAccount,
                @BsplType,
                @TransactionElementType,
                @ExpenseType,
                @DisplayOrder,
                @IsAggregationTarget,
                @TaxCode,
                @Balance,
                CURRENT_TIMESTAMP,
                CURRENT_TIMESTAMP
            )
            RETURNING ""勘定科目ID""
        ";

        return await connection.ExecuteScalarAsync<int>(sql, account);
    }

    public async Task<Account?> FindByCodeAsync(string accountCode)
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = $@"
            SELECT {SelectColumns} FROM ""勘定科目マスタ""
            WHERE ""勘定科目コード"" = @AccountCode
        ";

        return await connection.QuerySingleOrDefaultAsync<Account>(sql, new { AccountCode = accountCode });
    }

    public async Task<Account?> FindByIdAsync(int accountId)
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = $@"
            SELECT {SelectColumns} FROM ""勘定科目マスタ""
            WHERE ""勘定科目ID"" = @AccountId
        ";

        return await connection.QuerySingleOrDefaultAsync<Account>(sql, new { AccountId = accountId });
    }

    public async Task<IEnumerable<Account>> FindAllAsync()
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = $@"
            SELECT {SelectColumns} FROM ""勘定科目マスタ""
            ORDER BY ""勘定科目コード""
        ";

        return await connection.QueryAsync<Account>(sql);
    }

    public async Task<IEnumerable<Account>> FindByTypeAsync(string accountType)
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = $@"
            SELECT {SelectColumns} FROM ""勘定科目マスタ""
            WHERE ""勘定科目種別"" = @AccountType::account_type
            ORDER BY ""表示順序"", ""勘定科目コード""
        ";

        return await connection.QueryAsync<Account>(sql, new { AccountType = accountType });
    }

    public async Task<IEnumerable<Account>> FindSummaryAccountsAsync()
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = $@"
            SELECT {SelectColumns} FROM ""勘定科目マスタ""
            WHERE ""合計科目"" = true
            ORDER BY ""表示順序"", ""勘定科目コード""
        ";

        return await connection.QueryAsync<Account>(sql);
    }

    public async Task<IEnumerable<Account>> FindDetailAccountsAsync()
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = $@"
            SELECT {SelectColumns} FROM ""勘定科目マスタ""
            WHERE ""合計科目"" = false
            ORDER BY ""表示順序"", ""勘定科目コード""
        ";

        return await connection.QueryAsync<Account>(sql);
    }

    public async Task<int> UpdateAsync(Account account)
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = @"
            UPDATE ""勘定科目マスタ""
            SET ""勘定科目名"" = @AccountName,
                ""勘定科目カナ"" = @AccountNameKana,
                ""勘定科目種別"" = @AccountType::account_type,
                ""合計科目"" = @IsSummaryAccount,
                ""BSPL区分"" = @BsplType,
                ""取引要素区分"" = @TransactionElementType,
                ""費用区分"" = @ExpenseType,
                ""表示順序"" = @DisplayOrder,
                ""集計対象"" = @IsAggregationTarget,
                ""課税取引コード"" = @TaxCode,
                ""残高"" = @Balance,
                ""更新日時"" = CURRENT_TIMESTAMP
            WHERE ""勘定科目コード"" = @AccountCode
        ";

        return await connection.ExecuteAsync(sql, account);
    }

    public async Task<int> UpdateBalanceAsync(string accountCode, decimal balance)
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = @"
            UPDATE ""勘定科目マスタ""
            SET ""残高"" = @Balance,
                ""更新日時"" = CURRENT_TIMESTAMP
            WHERE ""勘定科目コード"" = @AccountCode
        ";

        return await connection.ExecuteAsync(sql, new { AccountCode = accountCode, Balance = balance });
    }

    public async Task<int> DeleteAsync(string accountCode)
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = @"
            DELETE FROM ""勘定科目マスタ""
            WHERE ""勘定科目コード"" = @AccountCode
        ";

        return await connection.ExecuteAsync(sql, new { AccountCode = accountCode });
    }
}

File: app/fsharp-ms/src/FinancialAccounting.Infrastructure/Persistence/Repositories/AccountRepository.fs
namespace FinancialAccounting.Infrastructure.Persistence.Repositories

open System
open System.Data
open System.Threading.Tasks
open Npgsql
open Dapper
open FinancialAccounting.Domain.Entities
open FinancialAccounting.Application.Ports.Out

/// <summary>
/// DB 行モデル（accounts テーブル）
/// </summary>
[<CLIMutable>]
type AccountRow = {
    account_id: int
    account_code: string
    account_name: string
    account_name_kana: string
    account_type: string
    is_summary_account: bool
    bs_pl_type: string
    transaction_element_type: string
    expense_type: string
    display_order: int
    is_aggregation_target: bool
    tax_code: string
    balance: decimal
    created_at: DateTime
    updated_at: DateTime
}

/// <summary>
/// 勘定科目リポジトリ実装（Dapper）
/// </summary>
type AccountRepository(connectionString: string) =

    let createConnection () : IDbConnection =
        new NpgsqlConnection(connectionString)

    let toAccount (row: AccountRow) : Account =
        let accountType =
            Account.stringToAccountType row.account_type
            |> Option.defaultValue Asset

        let bsPlType =
            Account.stringToBsPlType row.bs_pl_type
            |> Option.defaultValue BalanceSheet

        let transactionElementType =
            Account.stringToTransactionElementType row.transaction_element_type
            |> Option.defaultValue Debit

        {
            AccountId = Some row.account_id
            AccountCode = row.account_code
            AccountName = row.account_name
            AccountNameKana = row.account_name_kana |> Option.ofObj
            AccountType = accountType
            IsSummaryAccount = row.is_summary_account
            BsPlType = bsPlType
            TransactionElementType = transactionElementType
            ExpenseType = row.expense_type |> Option.ofObj
            DisplayOrder = row.display_order
            IsAggregationTarget = row.is_aggregation_target
            TaxCode = row.tax_code |> Option.ofObj
            Balance = row.balance
            CreatedAt = row.created_at
            UpdatedAt = row.updated_at
        }

    interface IAccountRepository with
        member _.SaveAsync(account: Account) =
            task {
                use connection = createConnection()
                connection.Open()

                let sql = """
                    INSERT INTO accounts (
                        account_code, account_name, account_name_kana, account_type,
                        is_summary_account, bs_pl_type, transaction_element_type,
                        expense_type, display_order, is_aggregation_target, tax_code, balance
                    ) VALUES (
                        @AccountCode, @AccountName, @AccountNameKana, @AccountType,
                        @IsSummaryAccount, @BsPlType, @TransactionElementType,
                        @ExpenseType, @DisplayOrder, @IsAggregationTarget, @TaxCode, @Balance
                    ) RETURNING account_id
                """

                let parameters = {|
                    AccountCode = account.AccountCode
                    AccountName = account.AccountName
                    AccountNameKana = account.AccountNameKana |> Option.toObj
                    AccountType = Account.accountTypeToString account.AccountType
                    IsSummaryAccount = account.IsSummaryAccount
                    BsPlType = Account.bsPlTypeToString account.BsPlType
                    TransactionElementType = Account.transactionElementTypeToString account.TransactionElementType
                    ExpenseType = account.ExpenseType |> Option.toObj
                    DisplayOrder = account.DisplayOrder
                    IsAggregationTarget = account.IsAggregationTarget
                    TaxCode = account.TaxCode |> Option.toObj
                    Balance = account.Balance
                |}

                let! accountId = connection.ExecuteScalarAsync<int>(sql, parameters)
                return { account with AccountId = Some accountId }
            }

        member _.GetByIdAsync(id: int) =
            task {
                use connection = createConnection()
                connection.Open()

                let sql = "SELECT * FROM accounts WHERE account_id = @Id"
                let! results = connection.QueryAsync<AccountRow>(sql, {| Id = id |})

                return
                    results
                    |> Seq.tryHead
                    |> Option.map toAccount
            }

        member _.GetByCodeAsync(code: string) =
            task {
                use connection = createConnection()
                connection.Open()

                let sql = "SELECT * FROM accounts WHERE account_code = @Code"
                let! results = connection.QueryAsync<AccountRow>(sql, {| Code = code |})

                return
                    results
                    |> Seq.tryHead
                    |> Option.map toAccount
            }

        member _.GetAllAsync() =
            task {
                use connection = createConnection()
                connection.Open()

                let sql = "SELECT * FROM accounts ORDER BY display_order, account_code"
                let! results = connection.QueryAsync<AccountRow>(sql)

                return
                    results
                    |> Seq.map toAccount
                    |> Seq.toList
            }

        member _.GetByTypeAsync(accountType: AccountType) =
            task {
                use connection = createConnection()
                connection.Open()

                let sql = "SELECT * FROM accounts WHERE account_type = @AccountType ORDER BY display_order, account_code"
                let! results = connection.QueryAsync<AccountRow>(sql, {| AccountType = Account.accountTypeToString accountType |})

                return
                    results
                    |> Seq.map toAccount
                    |> Seq.toList
            }

        member _.UpdateAsync(account: Account) =
            task {
                use connection = createConnection()
                connection.Open()

                let sql = """
                    UPDATE accounts SET
                        account_name = @AccountName,
                        account_name_kana = @AccountNameKana,
                        is_summary_account = @IsSummaryAccount,
                        expense_type = @ExpenseType,
                        display_order = @DisplayOrder,
                        is_aggregation_target = @IsAggregationTarget,
                        tax_code = @TaxCode,
                        updated_at = @UpdatedAt
                    WHERE account_id = @AccountId
                """

                let parameters = {|
                    AccountId = account.AccountId |> Option.defaultValue 0
                    AccountName = account.AccountName
                    AccountNameKana = account.AccountNameKana |> Option.toObj
                    IsSummaryAccount = account.IsSummaryAccount
                    ExpenseType = account.ExpenseType |> Option.toObj
                    DisplayOrder = account.DisplayOrder
                    IsAggregationTarget = account.IsAggregationTarget
                    TaxCode = account.TaxCode |> Option.toObj
                    UpdatedAt = DateTime.UtcNow
                |}

                let! _ = connection.ExecuteAsync(sql, parameters)
                return account
            }

        member _.DeleteAsync(id: int) =
            task {
                use connection = createConnection()
                connection.Open()

                let sql = "DELETE FROM accounts WHERE account_id = @Id"
                let! rowsAffected = connection.ExecuteAsync(sql, {| Id = id |})
                return rowsAffected > 0
            }
